pkg/apis/storage/types.go

```diff
@@ -672,3 +672,53 @@ type CSIStorageCapacityList struct {
 	// Items is the list of CSIStorageCapacity objects.
 	Items []CSIStorageCapacity
 }
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// VolumeAttributesClass represents a specification of mutable volume attributes
+// defined by the CSI driver. The class can be specified during dynamic provisioning
+// of PersistentVolumeClaims, and changed in the PersistentVolumeClaim spec after provisioning.
+type VolumeAttributesClass struct {
+	metav1.TypeMeta
+
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+
+	// Name of the CSI driver
+	// This field is immutable.
+	DriverName string
+
+	// parameters hold volume attributes defined by the CSI driver. These values
+	// are opaque to the Kubernetes and are passed directly to the CSI driver.
+	// The underlying storage provider supports changing these attributes on an
+	// existing volume, however the parameters field itself is immutable. To
+	// invoke a volume update, a new VolumeAttributesClass should be created with
+	// new parameters, and the PersistentVolumeClaim should be updated to reference
+	// the new VolumeAttributesClass.
+	//
+	// The keys cannot be empty, and the maximum number of parameters is 512, with
+	// a cumulative max size of 256K. If the CSI driver rejects invalid parameters,
+	// the target PersistentVolumeClaim will be set to an "Infeasible" state in the
+	// modifyVolumeStatus field.
+	// +optional
+	Parameters map[string]string
```

Comment:
based on [this discussion](https://github.com/kubernetes/enhancements/pull/4298#discussion_r1368102451), I want to make this parameters field become required and add new comment like:

```go
// This field is required and must contain at least one key/value pair.
```

What do you think? @msau42 @sunnylovestiramisu

Reply:
The problem is that for volume creation, it is not required. So it may not be appropriate to put it in the VAC object.

Or you need to say if it is for a modify volume call, "This field is required and must contain at least one key/value pair."

Reply:
Right now this API has a couple ways to have "no parameters":

- Empty string for `PersistentVolumeClaimSpec.VolumeAttributesClassName`
- `VolumeAttributesClass` with unset Parameters
- `VolumeAttributesClass` with empty Parameters map

It's not entirely clear if these are all exactly the same, or if there is some subtle difference between them.

If we can simplify the API such that there is only one way to have "no parameters", that would seem best. It would make the API more consistent, easier to understand usages and less prone to subtle and unexpected differences in behavior cause by usages that seem "conceptually same but slightly different in how they are expressed in the API".  Is that possible?

Note that, today, once `PersistentVolumeClaimSpec.VolumeAttributesClassName` is set to a non-empty value, it cannot be set to an empty string, so if we decide to require that Parameters be a non-empty map (at least one entry), we're also making it impossible to set the params to "no parameters" once set to some value. Is that what we want? Today it would be possible to create a `VolumeAttributesClass` with no parameters and set `PersistentVolumeClaimSpec.VolumeAttributesClassName` to the newly created `VolumeAttributesClass`.

Reply:
Agree, the way for someone to specify no parameters is to set their PVC.spec.vacName to "". So VAC.parameters should be required.

I think the awkward scenario though is the backwards compatible scenario for PVC.spec.vacName. We want nil to mean use the default VAC, however we don't want to change existing PVCs which will also have vacName as nil. We *could* have some conditional behavior like "we only set default VAC if PVC is unbound", which will change behavior for new PVCs but not existing ones, but it does mean "no vac" can have 2 ways to express it: 1) nil vacName when bound 2) empty string vacName

Reply:
updated